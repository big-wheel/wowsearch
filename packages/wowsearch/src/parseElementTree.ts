import * as uniq from 'lodash.uniq'
import * as visitTree from '@moyuyc/visit-tree'

import {
  Config,
  CrawlerConfig,
  Selectors,
  Selector,
  StrictSelector
} from './types/Config'
import selectVal, {
  normalizeSelector,
  selectAll,
  selectOne,
  transformVal
} from './selectVal'
import DocumentNode from './types/DocumentNode'
import LvlNode from './types/LvlNode'
import TextNode from './types/TextNode'

export function parseLvlTypeLevel(type: string = ''): number {
  type = type.trim()
  if (/^lvl([\d+])$/.test(type)) {
    return Number(RegExp.$1)
  }
  return null
}

export function isLvlType(type: string): boolean {
  return parseLvlTypeLevel(type) !== null
}

export const LVL_TYPES = [
  'lvl0',
  'lvl1',
  'lvl2',
  'lvl3',
  'lvl4',
  'lvl5',
  'lvl6'
]

function matchSelector(
  node: Element,
  selectorList: Selector[]
): { node: Element; selector: Selector; selectorIndex: number } | null {
  let selectorNode = node
  let selectorIndex = selectorList.findIndex(s => {
    s = normalizeSelector(s)
    if (s.type === 'css') {
      if (typeof node.matches === 'function' && node.matches(s.selector)) {
        selectorNode = node
        return true
      }
      selectorNode = selectOne(node, s)
      return !!selectorNode
    } else {
      selectorNode = selectOne(node, s)
      return !!selectorNode
    }
  })

  if (selectorIndex >= 0) {
    return {
      selector: selectorList[selectorIndex],
      selectorIndex,
      node: selectorNode
    }
  }

  return null
}

function getAnchor(elem: Element, selectorItem: StrictSelector) {
  const {
    anchor_selector = 'a[id]',
    anchor_attribute_name = 'id'
  } = selectorItem
  let anchor = elem.getAttribute(anchor_attribute_name)
  if (!anchor) {
    let tmpNode = selectOne(elem, anchor_selector)
    if (tmpNode) {
      return tmpNode.getAttribute(anchor_attribute_name)
    }
  }
  return anchor
}

function generateTextNode(node: Node, selector: StrictSelector) {
  return new TextNode(transformVal(node.textContent, selector))
}

// todo:
// 1. walk document by pre order
// 2. find node which matches one of selectors
function generateLvlNode(
  selectorKey,
  startElem: Element,
  selectors: Selectors,
  walk?: Function
): {lvlNode: LvlNode, endElement: Element} {
  const selectorItem = normalizeSelector(selectors[selectorKey])
  if (!selectorItem) return null
  const level = parseLvlTypeLevel(selectorKey)
  if (level === null) {
    // new TextNode(transformVal(current.textContent, selectorItem))
    return null
  }

  const selectorKeys = uniq(LVL_TYPES.concat(Object.keys(selectors)))
  const selectorEntryList = selectorKeys.map(key => ({key, selector: selectors[key]})).filter(({selector}) => selector)
  const selectorList = selectorEntryList.map(({selector}) => selector)
  const lvlNode = new LvlNode(level)

  walk && walk(startElem)
  lvlNode.value = transformVal(startElem.textContent, selectorItem)
  lvlNode.anchor = getAnchor(startElem, selectorItem)
  const children = lvlNode.children

  let callbackList = []
  let curr = startElem.nextElementSibling
  while (curr) {
    let isBreak
    let isContinue

    runUntilEq((current) => {
      let matches
      if ((matches = matchSelector(current, selectorList))) {
        const matchedSelectorKey = selectorEntryList[matches.selectorIndex] && selectorEntryList[matches.selectorIndex].key
        if (isLvlType(matchedSelectorKey)) {
          if (level < parseLvlTypeLevel(matchedSelectorKey)) {
            // callbackList.forEach(fn => fn())
            // callbackList = []

            const childObj = generateLvlNode(matchedSelectorKey, matches.node, selectors, walk)
            children.push(childObj.lvlNode)
            // matches.node.remove()
            curr = childObj.endElement
            isContinue = true
            return
            // return matches.node
          } else {
            isBreak = true
            return
            // return matches.node
          }
        } else if (matchedSelectorKey === 'text') {
          walk && walk(matches.node)
          children.push(generateTextNode(matches.node, matches.selector))
          callbackList.push(() => {
            matches.node.remove()
          })

        }

        return matches.node
      }
    }, curr)

    if (isBreak) {
      break
    }
    if (isContinue) {
      continue
    }

    // curr && curr.remove()
    curr = curr && curr.nextElementSibling
  }

  callbackList.forEach(fn => fn())

  return {
    lvlNode,
    endElement: curr
  }
}

function runUntilEq(walk, node: Element) {
  let next = walk(node)
  // node.remove && node.remove()

  while (next && next !== node) {
    next.remove()
    // node = next
    next = walk(node)
  }
}

export default function parseElementTree(
  document: Element,
  selectors: Selectors
): DocumentNode {
  const documentNode = new DocumentNode()
  const selectorKeys = uniq(LVL_TYPES.concat(Object.keys(selectors)))

  // Deals with global selector
  selectorKeys.forEach(selectorKey => {
    const selectorItem = normalizeSelector(selectors[selectorKey])
    if (!selectorItem) return

    if (selectorItem.global) {
      documentNode.global.set(
        selectorKey,
        selectVal(selectorItem, document).text
      )
    }
  })

  const selectorList = selectorKeys
    .map(key => ({
      key,
      selector: normalizeSelector(selectors[key])
    }))
    .filter(({ selector }) => selector && !selector.global)
  const selectorListPure = selectorList.map(({ selector }) => selector)

  const track = new WeakMap()
  visitTree(
    document,
    () => {},
    (childNode: Element, ctx) => {
      // Removed or walked
      if (track.has(childNode) || !childNode.parentElement) {
        return ctx.skip()
      }

      const matched = matchSelector(childNode, selectorListPure)
      if (matched) {
        if (matched.selector) {
          const matchedNode = matched.node
          if (track.has(matchedNode)) {
            return
          }

          const { selector, key } = selectorList[matched.selectorIndex]
          if (selector.type) {
            if (key === 'text') {
              documentNode.children.push(
                generateTextNode(matchedNode, selector)
              )
              // ctx.skip()
            } else if (isLvlType(key)) {
              const {lvlNode, endElement} = generateLvlNode(key, matchedNode, selectors, (elem) => {
                track.set(elem, true)
              })
              documentNode.children.push(lvlNode)
              // ctx.skip()
            }
          }
        }
        ctx.node.remove()
        return matched.node.remove()
      }
      else {
        ctx.skip()
        // childNode.remove && childNode.remove()
      }
    },
    { path: 'children' }
  )

  return documentNode
}
