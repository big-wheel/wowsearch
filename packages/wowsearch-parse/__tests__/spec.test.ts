/**
 * @file main.test
 * @author Cuttle Cong
 * @date 2018/6/10
 * @description
 */
import { readFileSync } from 'fs'
import { makeFixture } from './help'
import parseElementTree from '..'
import { JSDOM } from 'jsdom'

const parse = (html, config) => {
  const { document } = new JSDOM(html).window
  return parseElementTree(document.body, config)
}

// TODO performance
describe('parseElementTree', function() {
  it('spec wiki', async function () {
    const docNode = await parse(
      readFileSync(makeFixture('wiki.html')).toString(),
      {
        "author": {
          "global": true,
          "selector": ".page-metadata .author"
        },
        "lvl0": {
          "selector": "#breadcrumbs > li:last-child a",
          "global": true
        },
        "lvl1": ".wiki-content h1",
        "lvl2": ".wiki-content h2",
        "lvl3": ".wiki-content h3",
        "lvl4": ".wiki-content h4",
        "text": ".wiki-content p, .wiki-content li, .wiki-content pre"
      }
    )

    expect(docNode).toMatchSnapshot()
  });

  it('spec edam', async function() {
    const docNode = await parse(
      readFileSync(makeFixture('edam.html')).toString(),
      {
        lvl0: {
          selector:
            "//*[contains(@class,'navGroupActive')]//a[contains(@class,'navItemActive')]/preceding::h3[1]",
          type: 'xpath',
          global: true,
          default_value: 'Documentation'
        },
        lvl1: {
          type: 'css',
          global: true,
          selector: '.post h1'
        },
        lvl2: '.post article h1',
        lvl3: '.post article h2',
        lvl4: '.post article h3',
        lvl5: '.post article h4',
        text: '.post article p, .post article li'
      }
    )

    expect(docNode).toMatchSnapshot()
  })

  it('spec', async function() {
    const docNode = await parse(
      readFileSync(makeFixture('text.html')).toString(),
      {
        lvl1: '#readme h1:not(.package-name-redundant)',
        lvl0: {
          global: true,
          selector: '#readme .package-name-redundant',
          type: 'css'
        },
        lvl2: {
          selector: '//*[@id="readme"]//h2',
          type: 'xpath'
        },
        lvl3: {
          selector: '#readme h3'
        },
        lvl4: { selector: '#readme h4', default_value: 'abc' },
        text: '#readme > p, #readme li, #readme code, #readme pre'
      }
    )

    expect(docNode).toMatchSnapshot()
  })

  it('spec without global', async function() {
    const docNode = await parse(
      readFileSync(makeFixture('text.html')).toString(),
      {
        lvl1: '#readme h1:not(.package-name-redundant)',
        lvl0: '#readme .package-name-redundant',
        lvl2: {
          selector: '//*[@id="readme"]//h2',
          type: 'xpath'
        },
        lvl3: {
          selector: '#readme h3'
        },
        lvl4: { selector: '#readme h4', default_value: 'abc' },
        text: '#readme > p, #readme li, #readme code, #readme pre'
      }
    )

    expect(docNode).toMatchSnapshot()
  })

  it('should doc.html', async function () {
    const docNode = await parse(
      readFileSync(makeFixture('doc.html')).toString(),
      {
        lvl0: '.Post-RichText h1',
        lvl1: '.Post-RichText h2',
        lvl2: '.Post-RichText h3',
        lvl3: '.Post-RichText h4',
        lvl4: '.Post-RichText h5',
        lvl5: '.Post-RichText h6',
        code: '.Post-RichText pre',
        text:
          '.Post-RichText table, .Post-RichText p, .Post-RichText img, .Post-RichText ul, .Post-RichText ol, .Post-RichText li'
      }
    )
    expect(docNode).toMatchSnapshot()
  });
})
