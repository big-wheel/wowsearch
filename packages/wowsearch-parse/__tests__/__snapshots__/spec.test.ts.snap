// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`parseElementTree spec 1`] = `
DocumentNode {
  "children": Array [
    TextNode {
      "type": "text",
      "value": "robots.js — is parser for robots.txt files for node.js.",
    },
    LvlNode {
      "anchor": "user-content-installation",
      "children": Array [
        TextNode {
          "type": "text",
          "value": "It's recommended to install via npm:",
        },
        TextNode {
          "type": "text",
          "value": "$ npm install -g robots",
        },
      ],
      "level": 2,
      "type": "lvl",
      "value": "Installation",
    },
    LvlNode {
      "anchor": "user-content-usage",
      "children": Array [
        TextNode {
          "type": "text",
          "value": "Here's an example of using robots.js:",
        },
        TextNode {
          "type": "text",
          "value": "var robots = require('robots')  , parser = new robots.RobotsParser(); parser.setUrl('http://nodeguide.ru/robots.txt', function(parser, success) {  if(success) {    parser.canFetch('*', '/doc/dailyjs-nodepad/', function (access) {      if (access) {        // parse url      }    });  }});",
        },
        TextNode {
          "type": "text",
          "value": "Default crawler user-agent is:",
        },
        TextNode {
          "type": "text",
          "value": "Mozilla/5.0 (X11; Linux i686; rv:5.0) Gecko/20100101 Firefox/5.0
",
        },
        TextNode {
          "type": "text",
          "value": "Here's an example of using another user-agent and more detailed callback:",
        },
        TextNode {
          "type": "text",
          "value": "var robots = require('robots')  , parser = new robots.RobotsParser(                'http://nodeguide.ru/robots.txt',                'Mozilla/5.0 (compatible; RobotTxtBot/1.0)',                after_parse            );            function after_parse(parser, success) {  if(success) {    parser.canFetch('*', '/doc/dailyjs-nodepad/', function (access, url, reason) {      if (access) {        console.log(' url: '+url+', access: '+access);        // parse url ...      }    });  }};",
        },
        TextNode {
          "type": "text",
          "value": "Here's an example of getting list of sitemaps:",
        },
        TextNode {
          "type": "text",
          "value": "var robots = require('robots')  , parser = new robots.RobotsParser(); parser.setUrl('http://nodeguide.ru/robots.txt', function(parser, success) {  if(success) {    parser.getSitemaps(function(sitemaps) {      // sitemaps — array    });  }});",
        },
        TextNode {
          "type": "text",
          "value": "Here's an example of getCrawlDelay usage:",
        },
        TextNode {
          "type": "text",
          "value": "    var robots = require('robots')      , parser = new robots.RobotsParser();     // for example:    //    // $ curl -s http://nodeguide.ru/robots.txt    //    // User-agent: Google-bot    // Disallow: /     // Crawl-delay: 2    //    // User-agent: *    // Disallow: /    // Crawl-delay: 2     parser.setUrl('http://nodeguide.ru/robots.txt', function(parser, success) {      if(success) {        var GoogleBotDelay = parser.getCrawlDelay(\\"Google-bot\\");        // ...      }    });",
        },
        TextNode {
          "type": "text",
          "value": "An example of passing options to the HTTP request:",
        },
        TextNode {
          "type": "text",
          "value": "var options = {  headers:{    Authorization:\\"Basic \\" + new Buffer(\\"username:password\\").toString(\\"base64\\")}} var robots = require('robots')  , parser = new robots.RobotsParser(null, options); parser.setUrl('http://nodeguide.ru/robots.txt', function(parser, success) {  ...});",
        },
      ],
      "level": 2,
      "type": "lvl",
      "value": "Usage",
    },
    LvlNode {
      "anchor": "user-content-api",
      "children": Array [
        TextNode {
          "type": "text",
          "value": "RobotsParser — main class. This class provides a set of methods to read,
    parse and answer questions about a single robots.txt file.",
        },
        TextNode {
          "type": "text",
          "value": "setUrl(url, read) — sets the URL referring to a robots.txt file.
      by default, invokes read() method.
      If read is a function, it is called once the remote file is downloaded and parsed, and it
      takes in two arguments: the first is the parser itself, and the second is a boolean
      which is True if the the remote file was successfully parsed.",
        },
        TextNode {
          "type": "text",
          "value": "read(after_parse) — reads the robots.txt URL and feeds it to the parser",
        },
        TextNode {
          "type": "text",
          "value": "parse(lines) — parse the input lines from a robots.txt file",
        },
        TextNode {
          "type": "text",
          "value": "canFetch(userAgent, url, callback) — using the parsed robots.txt decide if
      userAgent can fetch url. Callback function:
      function callback(access, url, reason) { ... }
      where:
      
        access — can this url be fetched. true/false.
        url — target url
        reason — reason for access. Object:
          
            type — valid values: 'statusCode', 'entry', 'defaultEntry', 'noRule'
            entry — an instance of lib/Entry.js:. Only for types: 'entry', 'defaultEntry'
            statusCode — http response status code for url. Only for type 'statusCode'
          
        
      
    ",
        },
        TextNode {
          "type": "text",
          "value": "canFetchSync(userAgent, url) — using the parsed robots.txt decide if
      userAgent can fetch url. Return true/false.",
        },
        TextNode {
          "type": "text",
          "value": "getCrawlDelay(userAgent) — returns Crawl-delay for the certain userAgent",
        },
        TextNode {
          "type": "text",
          "value": "getSitemaps(sitemaps) — gets Sitemaps from parsed robots.txt",
        },
        TextNode {
          "type": "text",
          "value": "getDisallowedPaths(userAgent) — gets paths explictly disallowed for the user agent specified AND *",
        },
      ],
      "level": 2,
      "type": "lvl",
      "value": "API",
    },
    LvlNode {
      "anchor": "user-content-license",
      "children": Array [
        TextNode {
          "type": "text",
          "value": "See LICENSE
    file.",
        },
      ],
      "level": 2,
      "type": "lvl",
      "value": "License",
    },
    LvlNode {
      "anchor": "user-content-resources",
      "children": Array [
        TextNode {
          "type": "text",
          "value": "Robots.txt Specifications by Google",
        },
        TextNode {
          "type": "text",
          "value": "Robots.txt parser for python",
        },
        TextNode {
          "type": "text",
          "value": "A Standard for Robot Exclusion",
        },
      ],
      "level": 1,
      "type": "lvl",
      "value": "Resources",
    },
  ],
  "global": Map {
    "lvl0" => "robots.js",
  },
  "type": "document",
  "value": null,
}
`;

exports[`parseElementTree spec edam 1`] = `
DocumentNode {
  "children": Array [
    TextNode {
      "type": "text",
      "value": "Template is the most significant concept. the processing is follows.",
    },
    TextNode {
      "type": "text",
      "value": "Get user's answers by prompts.",
    },
    TextNode {
      "type": "text",
      "value": null,
    },
    TextNode {
      "type": "text",
      "value": "Read root files, filter the files by ignore.",
    },
    TextNode {
      "type": "text",
      "value": null,
    },
    TextNode {
      "type": "text",
      "value": "Transform file content using Loader",
    },
    TextNode {
      "type": "text",
      "value": "Rule one(top priority): matching the text from the file's first line.",
    },
    TextNode {
      "type": "text",
      "value": "Rule two(normal priority): matching loader from the test field of mappers.",
    },
    TextNode {
      "type": "text",
      "value": "// @loader \${LOADER_NAME}?\${QUERY}
/* @loader \${LOADER_NAME}?\${QUERY} */
# @loader \${LOADER_NAME}?\${QUERY}
    <!-- @loader \${LOADER_NAME}?\${QUERY} -->",
    },
    TextNode {
      "type": "text",
      "value": "Adjust files struction by copy and move.",
    },
    TextNode {
      "type": "text",
      "value": null,
    },
    TextNode {
      "type": "text",
      "value": "Trigger usefulHook then post hooks",
    },
    TextNode {
      "type": "text",
      "value": null,
    },
    TextNode {
      "type": "text",
      "value": "Note：The order of fields in package.json is that edam:main is superior to main. (>=3.1)",
    },
    LvlNode {
      "anchor": null,
      "children": Array [
        TextNode {
          "type": "text",
          "value": "Allows exporting config object directly, or callable.",
        },
        TextNode {
          "type": "text",
          "value": "module.exports = { /*config*/ } module.exports = edam => ({ /*config*/ })",
        },
        LvlNode {
          "anchor": null,
          "children": Array [
            TextNode {
              "type": "text",
              "value": "edam >= 2.0.1 Returns the object which excludes prompts.",
            },
            TextNode {
              "type": "text",
              "value": "type: function",
            },
            TextNode {
              "type": "text",
              "value": "Eg：",
            },
          ],
          "level": 4,
          "type": "lvl",
          "value": "process",
        },
        LvlNode {
          "anchor": null,
          "children": Array [
            TextNode {
              "type": "text",
              "value": "User's interaction definition.",
            },
            TextNode {
              "type": "text",
              "value": "type: []references to inquirer.js",
            },
          ],
          "level": 4,
          "type": "lvl",
          "value": "prompts",
        },
        LvlNode {
          "anchor": null,
          "children": Array [
            TextNode {
              "type": "text",
              "value": "Template files' root.",
            },
            TextNode {
              "type": "text",
              "value": "type: string",
            },
            TextNode {
              "type": "text",
              "value": "default: ./template",
            },
          ],
          "level": 4,
          "type": "lvl",
          "value": "root",
        },
        LvlNode {
          "anchor": null,
          "children": Array [
            TextNode {
              "type": "text",
              "value": "Files' filter",
            },
            TextNode {
              "type": "text",
              "value": "type: string[]Or answers => [",
            },
          ],
          "level": 4,
          "type": "lvl",
          "value": "ignore",
        },
        LvlNode {
          "anchor": null,
          "children": Array [
            TextNode {
              "type": "text",
              "value": "type {}",
            },
            TextNode {
              "type": "text",
              "value": "example{
    date: new Date(),
    val: 'abc'
}
    // or
(answers) => ({})",
            },
          ],
          "level": 4,
          "type": "lvl",
          "value": "variables",
        },
        LvlNode {
          "anchor": null,
          "children": Array [],
          "level": 4,
          "type": "lvl",
          "value": "hooks",
        },
        LvlNode {
          "anchor": null,
          "children": Array [
            TextNode {
              "type": "text",
              "value": "Check out Write a loader",
            },
          ],
          "level": 4,
          "type": "lvl",
          "value": "loaders 和 mappers",
        },
        LvlNode {
          "anchor": null,
          "children": Array [
            TextNode {
              "type": "text",
              "value": "or (answers) => ({})",
            },
          ],
          "level": 4,
          "type": "lvl",
          "value": "move",
        },
        LvlNode {
          "anchor": null,
          "children": Array [
            TextNode {
              "type": "text",
              "value": "like move",
            },
          ],
          "level": 4,
          "type": "lvl",
          "value": "copy",
        },
        LvlNode {
          "anchor": null,
          "children": Array [
            TextNode {
              "type": "text",
              "value": "Some useful shortcuts of post hook.",
            },
            TextNode {
              "type": "text",
              "value": "gitInitgit inittype: booleandefault: false",
            },
            TextNode {
              "type": "text",
              "value": "installDependenciesinstall dependencies of package.jsontype: booleandefault: false",
            },
            TextNode {
              "type": "text",
              "value": "installDevDependenciesinstall devDependencies of package.jsontype: booleandefault: false",
            },
            TextNode {
              "type": "text",
              "value": "或者 (answers) => ({})",
            },
          ],
          "level": 4,
          "type": "lvl",
          "value": "usefulHook",
        },
      ],
      "level": 3,
      "type": "lvl",
      "value": "Config",
    },
  ],
  "global": Map {
    "lvl0" => "Advanced",
    "lvl1" => "Write a template",
  },
  "type": "document",
  "value": null,
}
`;

exports[`parseElementTree spec without global 1`] = `
DocumentNode {
  "children": Array [
    LvlNode {
      "anchor": "user-content-robotsjs",
      "children": Array [
        TextNode {
          "type": "text",
          "value": "robots.js — is parser for robots.txt files for node.js.",
        },
        LvlNode {
          "anchor": "user-content-installation",
          "children": Array [
            TextNode {
              "type": "text",
              "value": "It's recommended to install via npm:",
            },
            TextNode {
              "type": "text",
              "value": "$ npm install -g robots",
            },
          ],
          "level": 2,
          "type": "lvl",
          "value": "Installation",
        },
        LvlNode {
          "anchor": "user-content-usage",
          "children": Array [
            TextNode {
              "type": "text",
              "value": "Here's an example of using robots.js:",
            },
            TextNode {
              "type": "text",
              "value": "var robots = require('robots')  , parser = new robots.RobotsParser(); parser.setUrl('http://nodeguide.ru/robots.txt', function(parser, success) {  if(success) {    parser.canFetch('*', '/doc/dailyjs-nodepad/', function (access) {      if (access) {        // parse url      }    });  }});",
            },
            TextNode {
              "type": "text",
              "value": "Default crawler user-agent is:",
            },
            TextNode {
              "type": "text",
              "value": "Mozilla/5.0 (X11; Linux i686; rv:5.0) Gecko/20100101 Firefox/5.0
",
            },
            TextNode {
              "type": "text",
              "value": "Here's an example of using another user-agent and more detailed callback:",
            },
            TextNode {
              "type": "text",
              "value": "var robots = require('robots')  , parser = new robots.RobotsParser(                'http://nodeguide.ru/robots.txt',                'Mozilla/5.0 (compatible; RobotTxtBot/1.0)',                after_parse            );            function after_parse(parser, success) {  if(success) {    parser.canFetch('*', '/doc/dailyjs-nodepad/', function (access, url, reason) {      if (access) {        console.log(' url: '+url+', access: '+access);        // parse url ...      }    });  }};",
            },
            TextNode {
              "type": "text",
              "value": "Here's an example of getting list of sitemaps:",
            },
            TextNode {
              "type": "text",
              "value": "var robots = require('robots')  , parser = new robots.RobotsParser(); parser.setUrl('http://nodeguide.ru/robots.txt', function(parser, success) {  if(success) {    parser.getSitemaps(function(sitemaps) {      // sitemaps — array    });  }});",
            },
            TextNode {
              "type": "text",
              "value": "Here's an example of getCrawlDelay usage:",
            },
            TextNode {
              "type": "text",
              "value": "    var robots = require('robots')      , parser = new robots.RobotsParser();     // for example:    //    // $ curl -s http://nodeguide.ru/robots.txt    //    // User-agent: Google-bot    // Disallow: /     // Crawl-delay: 2    //    // User-agent: *    // Disallow: /    // Crawl-delay: 2     parser.setUrl('http://nodeguide.ru/robots.txt', function(parser, success) {      if(success) {        var GoogleBotDelay = parser.getCrawlDelay(\\"Google-bot\\");        // ...      }    });",
            },
            TextNode {
              "type": "text",
              "value": "An example of passing options to the HTTP request:",
            },
            TextNode {
              "type": "text",
              "value": "var options = {  headers:{    Authorization:\\"Basic \\" + new Buffer(\\"username:password\\").toString(\\"base64\\")}} var robots = require('robots')  , parser = new robots.RobotsParser(null, options); parser.setUrl('http://nodeguide.ru/robots.txt', function(parser, success) {  ...});",
            },
          ],
          "level": 2,
          "type": "lvl",
          "value": "Usage",
        },
        LvlNode {
          "anchor": "user-content-api",
          "children": Array [
            TextNode {
              "type": "text",
              "value": "RobotsParser — main class. This class provides a set of methods to read,
    parse and answer questions about a single robots.txt file.",
            },
            TextNode {
              "type": "text",
              "value": "setUrl(url, read) — sets the URL referring to a robots.txt file.
      by default, invokes read() method.
      If read is a function, it is called once the remote file is downloaded and parsed, and it
      takes in two arguments: the first is the parser itself, and the second is a boolean
      which is True if the the remote file was successfully parsed.",
            },
            TextNode {
              "type": "text",
              "value": "read(after_parse) — reads the robots.txt URL and feeds it to the parser",
            },
            TextNode {
              "type": "text",
              "value": "parse(lines) — parse the input lines from a robots.txt file",
            },
            TextNode {
              "type": "text",
              "value": "canFetch(userAgent, url, callback) — using the parsed robots.txt decide if
      userAgent can fetch url. Callback function:
      function callback(access, url, reason) { ... }
      where:
      
        access — can this url be fetched. true/false.
        url — target url
        reason — reason for access. Object:
          
            type — valid values: 'statusCode', 'entry', 'defaultEntry', 'noRule'
            entry — an instance of lib/Entry.js:. Only for types: 'entry', 'defaultEntry'
            statusCode — http response status code for url. Only for type 'statusCode'
          
        
      
    ",
            },
            TextNode {
              "type": "text",
              "value": "canFetchSync(userAgent, url) — using the parsed robots.txt decide if
      userAgent can fetch url. Return true/false.",
            },
            TextNode {
              "type": "text",
              "value": "getCrawlDelay(userAgent) — returns Crawl-delay for the certain userAgent",
            },
            TextNode {
              "type": "text",
              "value": "getSitemaps(sitemaps) — gets Sitemaps from parsed robots.txt",
            },
            TextNode {
              "type": "text",
              "value": "getDisallowedPaths(userAgent) — gets paths explictly disallowed for the user agent specified AND *",
            },
          ],
          "level": 2,
          "type": "lvl",
          "value": "API",
        },
        LvlNode {
          "anchor": "user-content-license",
          "children": Array [
            TextNode {
              "type": "text",
              "value": "See LICENSE
    file.",
            },
          ],
          "level": 2,
          "type": "lvl",
          "value": "License",
        },
        LvlNode {
          "anchor": "user-content-resources",
          "children": Array [
            TextNode {
              "type": "text",
              "value": "Robots.txt Specifications by Google",
            },
            TextNode {
              "type": "text",
              "value": "Robots.txt parser for python",
            },
            TextNode {
              "type": "text",
              "value": "A Standard for Robot Exclusion",
            },
          ],
          "level": 1,
          "type": "lvl",
          "value": "Resources",
        },
      ],
      "level": 0,
      "type": "lvl",
      "value": "robots.js",
    },
  ],
  "global": Map {},
  "type": "document",
  "value": null,
}
`;
